File: cpp/src/ports.cpp
#include "tide/ports.h"
#include <stdexcept>

namespace tide {
namespace {
PortIndex index_ports(Index nodes, const std::vector<Index>& edges, const std::vector<Index>& ports,
                      const std::vector<Index>& edge_slots, const std::vector<Index>& port_slots) {
  if (edges.size() != edge_slots.size() || ports.size() != port_slots.size())
    throw std::invalid_argument("local slot layout size mismatch");
  PortIndex result;
  result.offsets.assign(nodes + 1, 0);
  for (auto node : edges) ++result.offsets[node + 1];
  for (auto node : ports) ++result.offsets[node + 1];
  for (Index v = 0; v < nodes; ++v) result.offsets[v + 1] += result.offsets[v];
  result.bindings.assign(edges.size() + ports.size(), {-1, -1});
  auto assign = [&](const auto& owners, const auto& slots, Index kind) {
    for (size_t id = 0; id < owners.size(); ++id) {
      const auto node = owners[id], slot = slots[id];
      const auto start = result.offsets[node], count = result.offsets[node + 1] - start;
      if (slot < 0 || slot >= count || result.bindings[start + slot].kind != -1)
        throw std::invalid_argument("local slots must be a bijection for each node/direction");
      result.bindings[start + slot] = {kind, static_cast<Index>(id)};
    }
  };
  assign(edges, edge_slots, 1); assign(ports, port_slots, 0);
  return result;
}
}  // namespace
void compile_ports(Graph& g) {
  std::vector<Index> source, target;
  for (const auto& e : g.edges) { source.push_back(e.source); target.push_back(e.target); }
  if (!g.layout) {
    PortLayout layout;
    std::vector<Index> incoming(g.nodes.size(), 0), outgoing(g.nodes.size(), 0);
    auto allocate = [](const auto& owners, auto& counts) {
      std::vector<Index> slots;
      for (auto node : owners) slots.push_back(counts[node]++);
      return slots;
    };
    layout.input = allocate(g.inputs, incoming);
    layout.edge_target = allocate(target, incoming);
    layout.edge_source = allocate(source, outgoing);
    layout.output = allocate(g.outputs, outgoing);
    g.layout = std::move(layout);
  }
  const auto& layout = *g.layout;
  g.incoming_ports = index_ports(g.nodes.size(), target, g.inputs, layout.edge_target, layout.input);
  g.outgoing_ports = index_ports(g.nodes.size(), source, g.outputs, layout.edge_source, layout.output);
}
}  // namespace tide

File: python/tidegraph/ports.py
"""Stable local program slots, independent of physical edge/port identities."""
from dataclasses import dataclass


@dataclass(frozen=True)
class PortIndex:
    offsets: tuple[int, ...]
    bindings: tuple[tuple[int, int], ...]  # kind 0: boundary port; kind 1: edge

    def row(self, node):
        return self.bindings[self.offsets[node]:self.offsets[node + 1]]


@dataclass(frozen=True)
class PortLayout:
    edge_source: tuple[int, ...]
    edge_target: tuple[int, ...]
    input: tuple[int, ...]
    output: tuple[int, ...]

    def __post_init__(self):
        for name in ("edge_source", "edge_target", "input", "output"):
            values = tuple(getattr(self, name))
            if any(type(x) is not int or not 0 <= x < 2**63 for x in values):
                raise ValueError("local slots require nonnegative int64 values")
            object.__setattr__(self, name, values)

    @classmethod
    def automatic(cls, graph):
        incoming, outgoing = [0] * len(graph.nodes), [0] * len(graph.nodes)
        def allocate(owners, counts):
            slots = []
            for node in owners:
                slots.append(counts[node]); counts[node] += 1
            return tuple(slots)
        # Incoming boundaries precede edges, matching the default atom order.
        inputs = allocate(graph.inputs, incoming)
        targets = allocate((e.target for e in graph.edges), incoming)
        sources = allocate((e.source for e in graph.edges), outgoing)
        outputs = allocate(graph.outputs, outgoing)
        return cls(sources, targets, inputs, outputs)

    def indexes(self, graph):
        """Validate a bijection per direction/node; compile flat slot->wire rows."""
        def index(edge_owners, ports, edge_slots, port_slots):
            if len(edge_slots) != len(edge_owners) or len(port_slots) != len(ports):
                raise ValueError("local slot layout size mismatch")
            rows = [[] for _ in graph.nodes]
            for node in (*edge_owners, *ports):
                rows[node].append(None)
            for kind, owners, slots in ((1, edge_owners, edge_slots), (0, ports, port_slots)):
                for source, (node, slot) in enumerate(zip(owners, slots)):
                    if slot >= len(rows[node]) or rows[node][slot] is not None:
                        raise ValueError("local slots must be a bijection for each node/direction")
                    rows[node][slot] = kind, source
            offsets, bindings = [0], []
            for row in rows:
                bindings.extend(row); offsets.append(len(bindings))
            return PortIndex(tuple(offsets), tuple(bindings))
        incoming = index(tuple(e.target for e in graph.edges), graph.inputs, self.edge_target, self.input)
        outgoing = index(tuple(e.source for e in graph.edges), graph.outputs, self.edge_source, self.output)
        return incoming, outgoing

    def incoming_slot(self, kind, source):
        if kind not in (0, 1):
            raise ValueError("invalid source kind")
        return (self.input if kind == 0 else self.edge_target)[source]
